hoomd/mpcd/ReverseNonequilibriumShearFlow.cc

```diff
@@ -0,0 +1,377 @@
+    m_particles_staged.swap(swap_queue);
+    }
+
+mpcd::ReverseNonequilibriumShearFlow::~ReverseNonequilibriumShearFlow()
+    {
+    m_exec_conf->msg->notice(5) << "Destroying ReverseNonequilibriumShearFlow" << std::endl;
+    m_pdata->getBoxChangeSignal()
+        .disconnect<mpcd::ReverseNonequilibriumShearFlow,
+                    &mpcd::ReverseNonequilibriumShearFlow::requestUpdateSlabs>(this);
+    }
+
+/*!
+ * \param num_swap Max number of swaps
+ */
+void mpcd::ReverseNonequilibriumShearFlow::setNumSwap(unsigned int num_swap)
+    {
+    if (num_swap > m_num_swap)
+        {
+        GPUArray<Scalar2> swap_queue(2 * num_swap, m_exec_conf);
+        m_particles_staged.swap(swap_queue);
+        }
+    m_num_swap = num_swap;
+    }
+
+//! Set the target momentum
+void mpcd::ReverseNonequilibriumShearFlow::setTargetMomentum(Scalar target_momentum)
+    {
+    m_target_momentum = target_momentum;
+    }
+
+/*!
+ * \param slab_width Slab thickness
+ */
+void mpcd::ReverseNonequilibriumShearFlow::setSlabWidth(Scalar slab_width)
+    {
+    m_slab_width = slab_width;
+    requestUpdateSlabs();
+    }
+
+void mpcd::ReverseNonequilibriumShearFlow::setSlabs()
+    {
+    const BoxDim& global_box = m_pdata->getGlobalBox();
+    if (m_slab_width > Scalar(0.5) * global_box.getL().y)
+        {
+        throw std::runtime_error("Slab thickness cannot be larger than Ly/2");
+        }
+
+    const Scalar3 global_lo = global_box.getLo();
+    m_pos_lo = make_scalar2(global_lo.y, global_lo.y + m_slab_width);
+    m_pos_hi = make_scalar2(Scalar(0.0), m_slab_width);
+    }
+
+/*!
+ * \param timestep Current time step of the simulation
+ */
+void mpcd::ReverseNonequilibriumShearFlow::update(uint64_t timestep)
+    {
+    // if slabs have changed, update them
+    if (m_update_slabs)
+        {
+        setSlabs();
+        m_update_slabs = false;
+        }
+
+    findSwapParticles();
+    stageSwapParticles();
+    swapParticleMomentum();
+    }
+
+/*!
+ * Finds all particles in the "lower" and "upper" slab in y direction and
+ * puts them into two GPUArrays, sorted by their momentum closest to -/+ target_momentum in x
+ * direction.
+ */
+void mpcd::ReverseNonequilibriumShearFlow::findSwapParticles()
+    {
+    // fill the layers with (unsorted) particles. this uses do loop for reallocation
+    bool filled = false;
+    do
+        {
+        const size_t num_layer_lo_alloc = m_particles_lo.getNumElements();
+        const size_t num_layer_hi_alloc = m_particles_hi.getNumElements();
+            {
+            ArrayHandle<Scalar4> h_pos(m_mpcd_pdata->getPositions(),
+                                       access_location::host,
+                                       access_mode::read);
+            ArrayHandle<Scalar4> h_vel(m_mpcd_pdata->getVelocities(),
+                                       access_location::host,
+                                       access_mode::read);
+
+            ArrayHandle<Scalar2> h_layer_lo(m_particles_lo,
```

Comment:
Array handles should match GPU array names.
```suggestion
            ArrayHandle<Scalar2> h_particles_lo(m_particles_lo,
```